### PR TITLE
fix bf16 model type issue for ia3

### PR DESCRIPTION
**Pull request by sywangyi**

### CODE REVIEW

**Comment by sywangyi** [comment]
@BenjaminBossan please help review.

**Comment by sywangyi** [comment]
if set torch_dtype is bfloat16 in model.from_pretained(), all adapter layers after get_peft_model(xxx) should be bfloat16 as well. fix issue in ia3 and adalora, also ia3 forward return dtype should be aligned with the base_layer.forward() output.

**Review by BenjaminBossan — APPROVED** [review]
Thanks, those changes to correctly enable bf16 look good. I'll request another review by the co-maintainers just to be sure.

File: src/peft/tuners/adalora/layer.py
Comment (pacman100) [inline src/peft/tuners/adalora/layer.py:74]: this would cast the trainable lora weights too in half precision if `torch_dtype` is a half precision which is not good for stable training. We are discussion this internally as this has been used at other places too.

Comment (sywangyi) [reply src/peft/tuners/adalora/layer.py:74]: so I would update the PR and focus on IA3 forward return type bug. also IA3 merge weight has type issue in torch_dtype half precision case. if IA3's weight is float32 and base layer weight is bfloat16, after weight merged, it should be bfloat16 instead of float32, am I right?

**Comment by BenjaminBossan** [comment]
@pacman100 Could you please check again?

**Review by pacman100 — APPROVED** [review]
Thank you, @sywangyi, for the fixes!

**Pull request merged**
